File: misc/anagram/groupanagrams/GroupAnagrams.py
from collections import defaultdict

class GroupAnagrams:
    def group_anagrams(self, words):
        anagram_groups = defaultdict(list)

        for word in words:
            # Sort the characters of the word to use as a key
            sorted_word = ''.join(sorted(word))
            # Append the original word to the corresponding anagram group
            anagram_groups[sorted_word].append(word)

        return list(anagram_groups.values())

if __name__ == "__main__":
    anagram_grouper = GroupAnagrams()
    input_words = ["eat", "tea", "tan", "ate", "nat", "bat"]
    result = anagram_grouper.group_anagrams(input_words)
    print(result)
    # Output: [['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]

File: misc/anagram/groupanagrams/GroupAnagrams.java
package anagram.groupanagrams;

import java.util.*;

class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] words) {
        Map<String, List<String>> anagramGroups = new HashMap<>();

        for (String word : words) {
            char[] characters = word.toCharArray();
            Arrays.sort(characters);
            String sortedWord = new String(characters);

            // If the sorted string is not in the map, add it
            // Add the original string to the list corresponding to the sorted key
            anagramGroups.computeIfAbsent(sortedWord, k -> new ArrayList<>()).add(word);

            //OR
            /*if (!anagramGroups.containsKey(sortedWord)) {
                anagramGroups.put(sortedWord, new ArrayList<>());
            }
            anagramGroups.get(sortedWord).add(word);*/
        }

        return new ArrayList<>(anagramGroups.values());
    }

    public static void main(String[] args) {
        GroupAnagrams anagramGrouper = new GroupAnagrams();
        String[] inputWords = {"eat", "tea", "tan", "ate", "nat", "bat"};
        List<List<String>> result = anagramGrouper.groupAnagrams(inputWords);
        System.out.println(result);
        // Output: [[bat], [nat, tan], [ate, eat, tea]]
    }
}

